src/Common/src/System/IO/PersistedFiles.Unix.cs

```diff
@@ -0,0 +1,88 @@
+        private const string TopLevelUserDirectory = "." + TopLevelDirectory;
+        private const string SecondLevelDirectory = "corefx";
+
+        // Do not use the user variables in the static initializer in case $HOME didn't resolve.
+        private static string s_userHomeDirectory;
+        private static string s_userProductDirectory;
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of the framework,
+        /// such as "cryptography".
+        /// </summary>
+        /// <param name="featureName">The directory name for the feature</param>
+        /// <returns>A path within the user's home directory for persisting data for the feature</returns>
+        internal static string GetUserFeatureDirectory(string featureName)
+        {
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, featureName);
+        }
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of a feature of
+        /// the framework, such as "x509stores" within "cryptography".
+        /// </summary>
+        /// <param name="featureName">The directory name for the feature</param>
+        /// <param name="subFeatureName">The directory name for the sub-feature</param>
+        /// <returns>A path within the user's home directory for persisting data for the sub-feature</returns>
+        internal static string GetUserFeatureDirectory(string featureName, string subFeatureName)
+        {
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, featureName, subFeatureName);
+        }
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of the framework,
+        /// with a lot of hierarchy, such as ["cryptography", "x509stores", "my"]
+        /// </summary>
+        /// <param name="featurePathParts">A non-empty set of directories to use for the storage hierarchy</param>
+        /// <returns>A path within the user's home directory for persisting data for the feature</returns>
+        internal static string GetUserFeatureDirectory(params string[] featurePathParts)
+        {
+            Debug.Assert(featurePathParts != null);
+            Debug.Assert(featurePathParts.Length > 0);
+
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, Path.Combine(featurePathParts));
+        }
+
+        private static void EnsureUserDirectories()
+        {
+            // TODO (2820): Use getpwuid_r(geteuid(), ...) instead of the HOME environment variable
+            s_userHomeDirectory = Environment.GetEnvironmentVariable("HOME");
```

Comment:
It seems the only place this is needed is in subsequently computing s_userProductDirectory.  Could it be a local instead of a field?